File: dsa-problems/StringsClassic/solutions/cpp/STC-006-lcp-array-kasai.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> lcpArray(const string& s, const vector<int>& sa) {
        int n = s.length();
        vector<int> rank(n);
        for (int i = 0; i < n; i++) {
            rank[sa[i]] = i;
        }
        
        vector<int> lcp(n - 1);
        int k = 0;
        
        for (int i = 0; i < n; i++) {
            if (rank[i] == n - 1) {
                k = 0;
                continue;
            }
            
            int j = sa[rank[i] + 1];
            while (i + k < n && j + k < n && s[i + k] == s[j + k]) {
                k++;
            }
            
            lcp[rank[i]] = k;
            if (k > 0) k--;
        }
        return lcp;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    string s;
    if (cin >> s) {
        int n;
        cin >> n;
        vector<int> sa(n);
        for (int i = 0; i < n; i++) cin >> sa[i];
        
        Solution solution;
        vector<int> lcp = solution.lcpArray(s, sa);
        for (int i = 0; i < (int)lcp.size(); i++) {
            if (i > 0) cout << " ";
            cout << lcp[i];
        }
        cout << "\n";
    }
    return 0;
}

File: dsa-problems/StringsClassic/solutions/javascript/STC-006-lcp-array-kasai.js
const readline = require("readline");

class Solution {
  lcpArray(s, sa) {
    const n = s.length;
    const rank = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
      rank[sa[i]] = i;
    }
    
    const lcp = new Array(n - 1).fill(0);
    let k = 0;
    
    for (let i = 0; i < n; i++) {
      if (rank[i] === n - 1) {
        k = 0;
        continue;
      }
      
      const j = sa[rank[i] + 1];
      while (i + k < n && j + k < n && s[i + k] === s[j + k]) {
        k++;
      }
      
      lcp[rank[i]] = k;
      if (k > 0) k--;
    }
    return lcp;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => {
  const parts = line.trim().split(/\s+/);
  for (const part of parts) {
    if (part) data.push(part);
  }
});

rl.on("close", () => {
  if (data.length < 2) return;
  let idx = 0;
  const s = data[idx++];
  const n = parseInt(data[idx++], 10);
  if (isNaN(n) || idx + n > data.length) return;
  const sa = [];
  for (let i = 0; i < n; i++) sa.push(parseInt(data[idx++], 10));

  const solution = new Solution();
  const lcp = solution.lcpArray(s, sa);
  console.log(lcp.join(" "));
});
